src/NuGet.Core/NuGet.Commands/RestoreCommand/Logging/TransitiveNoWarnUtils.cs

```diff
@@ -720,37 +689,44 @@ private class LookUpNode
 
             // If a node is a project then it will hold these properties
             public NodeWarningProperties NodeWarningProperties { get; set; }
-
         }
 
-
         /// <summary>
-        /// A class to hold minimal version of project wide nowarn and package specific no warn for a project.
+        /// An immutable class to hold minimal version of project wide nowarn and package specific no warn for a project.
         /// </summary>
         public class NodeWarningProperties : IEquatable<NodeWarningProperties>
         {
+            // Empty NodeWarningProperties singleton - to avoid allocation for most-common case
+            internal static readonly NodeWarningProperties Empty = new NodeWarningProperties(null, null);
+
             // ProjectWide NoWarn properties
-            public HashSet<NuGetLogCode> ProjectWide { get; }
+            private readonly HashSet<NuGetLogCode> _projectWide;
```

Comment:
Why are we doing this refactoring?
I'm not sure what the functional benefit here is. 

Making a lot of changes makes it really difficult to keep up with the actual relevant changes.

Reply:
I apologize that the review has grown uglier.  My goal here was to emphasize that this HashSet and the below Dictionary are intended to be immutable.  Note that the "{ get; }" syntax is a shorthand for a getter property with a backing member variable that can only be set by the constructor.  This is basically equivalent to a private readonly variable with a getter property that returns it.

Originally, I had planned to remove the getter property altogether.  Unfortunately, this object had made it into the publicly-exposed API for NuGet, so we cannot make that change without a major version update.  I decided to keep it with the separated readonly member variable to hopefully help remind the next updater that this is something that can and should be removed at the next major version update.

I can revert this if you prefer and go back to the old syntax with a comment.

Reply:
As you mentioned, they are semantically equivalent, so it's probably easier if we didn't make those changes.